### PR TITLE
MiMo-V2-Flash Support

### PR DESCRIPTION
**Pull request by wufann**
## Motivation
support MiMo-V2-Flash model

## Technical Details
  1. Base Model (atom/models/mimo_v2_flash.py)
  - Implement full MiMo-V2-Flash inference architecture
  - Hybrid attention: Full Attention + SWA (Sliding Window Attention) selected per-layer via hybrid_layer_pattern, with different num_kv_heads per layer type (4 vs 8)
  - Register MiMoV2FlashForCausalLM in support_model_arch_dict

  2. Per-layer KV Cache Allocation (atom/model_engine/model_runner.py)
https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash/blob/main/config.json
"num_key_value_heads": 4
“swa_num_key_value_heads": 8,

  - MiMo-V2-Flash has varying num_kv_heads across layers, preventing a single unified KV cache tensor
  - Add is_mimo_v2() helper method
  - Memory estimation (_compute_block_bytes): use max(full_kv_heads, swa_kv_heads) as a conservative upper bound
  - KV cache allocation (_allocate_kv_cache): skip unified tensor pre-allocation, use _kv_layer_cache_store list instead
  - KV cache binding: allocate per-layer via torch.zeros during module traversal, sized by each module.num_kv_heads

  3. QKV Linear with Asymmetric head_dim (atom/model_ops/linear.py)

  - Add v_head_size parameter to QKVParallelLinear to support different K/V head dimensions (MiMo-V2: head_dim=192, v_head_dim=128)  
  - Fix FP8 block-wise quantization weight_scale shard calculation: //128 → (+127)//128 (ceil division) to handle non-128-aligned dimensions ref:https://github.com/vllm-project/vllm/blob/v0.19.0/vllm/model_executor/layers/linear.py#L92

  4. MHA FP8 KV Cache Write Fix (atom/model_ops/attention_mha.py)

  - For use_triton_attn (pa_decode_gluon) path, use per-tensor kv_scale with reshape_and_cache instead of per-token quant, matching the decode kernel's kv_quant_mode=0 expectation

  5. FP8 Quantization Config Parsing (atom/quant_spec.py)

  - Add ignored_layers key parsing to support MiMo-V2-Flash's HF quantization config format (excludes attention output projection layers from quantization)
  [o_proj dtype bf16](https://huggingface.co/XiaomiMiMo/MiMo-V2-Flash/blob/main/config.json#L15)

  6. MTP Speculative Decoding (atom/models/mimo_v2_flash_mtp.py, atom/spec_decode/eagle.py)

  - Implement MiMoV2FlashMTP draft model: 3 independent MTP layers (each with SWA attention + dense MLP), stored in ModuleDict and selected via spec_step_idx % num_mtp_layers
  - Weight remapping: model.mtp.layers.{N}.* → model.layers.{48+N}.*, distinguishing predictor-level weights (enorm/hnorm/eh_proj/final_layernorm) from transformer block weights (self_attn/mlp)
  - Pass spec_step_idx=i to draft model's forward() and compute_logits() in eagle.py
  - MHA attention metadata in propose(): branch on MLA vs MHA — MHA skips kv_last_page_lens, skips kv_indptr subtraction (block-level, not token-level), and increments context_lens instead

  7. MHA prepare_mtp_decode (atom/model_ops/attentions/aiter_attention.py)

  - Add prepare_mtp_decode method to AiterAttentionMetadataBuilder for updating kv_indices and persistent worker buffers between draft steps in MHA attention

  8. Config & Loader (atom/config.py, atom/model_loader/loader.py)

  - Register mimo_v2_flash → mimo_v2_flash_mtp model type mapping in SpeculativeConfig, default num_nextn_predict_layers to 3 (3 independent MTP layers, not reused)
  - Add trust_remote_code=True to AutoConfig.from_pretrained in SpeculativeConfig.__post_init__
  - Add mimo_v2_flash_mtp weight loading branch in loader.py

### Acc Perf
TP4 + FP8KV
TP4 + FP8KV + MTP
```
python -m atom.entrypoints.openai_server --model /data/MiMo-V2-Flash -tp 4 --trust-remote-code --kv_cache_dtype fp8

python -m atom.entrypoints.openai_server --model /data/MiMo-V2-Flash -tp 4 --trust-remote-code --kv_cache_dtype fp8 --method mtp

python -m atom.benchmarks.benchmark_serving \
  --model=/data/MiMo-V2-Flash --backend=vllm --base-url=http://localhost:8000 \
  --dataset-name=random \
  --random-input-len=1024 --random-output-len=1024 \
  --random-range-ratio=0.8 \
  --num-prompts=1280 --max-concurrency=128 \
  --request-rate=inf --ignore-eos \
  --save-result --percentile-metrics="ttft,tpot,itl,e2el"

lm_eval --model local-completions \
  --model_args model=/data/MiMo-V2-Flash,base_url=http://localhost:8000/v1/completions,num_concurrent=64,max_retries=3,tokenized_requests=False \
  --tasks gsm8k --num_fewshot 5
```

### Acc
TP4 + FP8KV
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8385|±  |0.0101|
|     |       |strict-match    |     5|exact_match|↑  |0.8324|±  |0.0103|

TP4 + FP8KV + MTP
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.840|±  |0.0101|
|     |       |strict-match    |     5|exact_match|↑  |0.837|±  |0.0102|

### Perf
TP4 + FP8KV
<details>
<summary>data</summary>

```bash
============ Serving Benchmark Result ============
Successful requests:                     1280
Benchmark duration (s):                  212.87
Total input tokens:                      1180188
Total generated tokens:                  1177601
Request throughput (req/s):              6.01
Output token throughput (tok/s):         5532.05
Total Token throughput (tok/s):          11076.26
---------------Time to First Token----------------
Mean TTFT (ms):                          272.98
Median TTFT (ms):                        94.78
P99 TTFT (ms):                           2678.93
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          22.25
Median TPOT (ms):                        22.58
P99 TPOT (ms):                           23.74
---------------Inter-token Latency----------------
Mean ITL (ms):                           22.24
Median ITL (ms):                         18.53
P99 ITL (ms):                            87.42
----------------End-to-end Latency----------------
Mean E2EL (ms):                          20731.09
Median E2EL (ms):                        20679.48
P99 E2EL (ms):                           24581.77
==================================================
```
</details>

TP4 + FP8KV + MTP
<details>
<summary>data</summary>

```bash
============ Serving Benchmark Result ============
Successful requests:                     1280
Benchmark duration (s):                  166.52
Total input tokens:                      1180188
Total generated tokens:                  1176670
Request throughput (req/s):              7.69
Output token throughput (tok/s):         7066.17
Total Token throughput (tok/s):          14153.46
---------------Time to First Token----------------
Mean TTFT (ms):                          248.35
Median TTFT (ms):                        106.24
P99 TTFT (ms):                           2466.34
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          17.31
Median TPOT (ms):                        16.68
P99 TPOT (ms):                           22.88
---------------Inter-token Latency----------------
Mean ITL (ms):                           30.34
Median ITL (ms):                         22.38
P99 ITL (ms):                            106.09
----------------End-to-end Latency----------------
Mean E2EL (ms):                          16137.63
Median E2EL (ms):                        15892.93
P99 E2EL (ms):                           21989.89
==================================================

```
</details>

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

cc: @billishyahao

### CODE REVIEW

**Comment by valarLip** [comment]
any mtp acceptance rate numbers？

**Comment by wufann** [comment]
> any mtp acceptance rate numbers？

can't see summary of the acceptance rate, here are some server log when run gsm8k dataset:
```
14278:[atom 14:45:50] [MTP Stats Interval] Average toks/fwd: 1.73, Accepted/Total Draft tokens: 727/1000, Acceptance rate: 72.70%, Accepted tokens distribution: {0: '27.30%', 1: '72.70%'}
14279:[atom 14:45:50] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 64982/93000, Acceptance rate: 69.87%, Accepted tokens distribution: {0: '30.13%', 1: '69.87%'}
14513:[atom 14:45:51] [MTP Stats Interval] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 703/1000, Acceptance rate: 70.30%, Accepted tokens distribution: {0: '29.70%', 1: '70.30%'}
14514:[atom 14:45:51] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 65685/94000, Acceptance rate: 69.88%, Accepted tokens distribution: {0: '30.12%', 1: '69.88%'}
14617:[atom 14:45:52] [MTP Stats Interval] Average toks/fwd: 1.73, Accepted/Total Draft tokens: 729/1000, Acceptance rate: 72.90%, Accepted tokens distribution: {0: '27.10%', 1: '72.90%'}
14618:[atom 14:45:52] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 66414/95000, Acceptance rate: 69.91%, Accepted tokens distribution: {0: '30.09%', 1: '69.91%'}
14771:[atom 14:45:53] [MTP Stats Interval] Average toks/fwd: 1.71, Accepted/Total Draft tokens: 705/1000, Acceptance rate: 70.50%, Accepted tokens distribution: {0: '29.50%', 1: '70.50%'}
14772:[atom 14:45:53] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 67119/96000, Acceptance rate: 69.92%, Accepted tokens distribution: {0: '30.08%', 1: '69.92%'}
14913:[atom 14:45:53] [MTP Stats Interval] Average toks/fwd: 1.72, Accepted/Total Draft tokens: 724/1000, Acceptance rate: 72.40%, Accepted tokens distribution: {0: '27.60%', 1: '72.40%'}
14914:[atom 14:45:53] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 67843/97000, Acceptance rate: 69.94%, Accepted tokens distribution: {0: '30.06%', 1: '69.94%'}
15022:[atom 14:45:54] [MTP Stats Interval] Average toks/fwd: 1.68, Accepted/Total Draft tokens: 682/1000, Acceptance rate: 68.20%, Accepted tokens distribution: {0: '31.80%', 1: '68.20%'}
15023:[atom 14:45:54] [MTP Stats         ] Average toks/fwd: 1.70, Accepted/Total Draft tokens: 68525/98000, Acceptance rate: 69.92%, Accepted tokens distribution: {0: '30.08%', 1: '69.92%'}
```

**Comment by billishyahao** [comment]
Some CI is failing. https://github.com/ROCm/ATOM/actions/runs/24436188452/job/71391165056?pr=560 
But seems to be OOM issue. may not related to this PR.

File: atom/model_engine/model_runner.py [hunk 1]
```diff
@@ -1079,6 +1087,32 @@ def _compute_block_bytes(self):
...
+            # MiMo-V2-Flash has mixed attention types (full + SWA) with
+            # different num_kv_heads per layer. Use the max across layer
+            # types as a conservative upper bound for memory estimation.
+            _swa_raw = getattr(hf_config, "swa_num_key_value_heads", 0)
+            _swa_per_rank = (
+                _swa_raw // self.world_size
+                if _swa_raw >= self.world_size
+                else (1 if _swa_raw else 0)
+            )
+            max_kv_heads = max(num_kv_heads, _swa_per_rank)
```
Comment (valarLip) [inline atom/model_engine/model_runner.py]: can we calculate here more accuracy?

File: atom/model_ops/attention_mha.py [hunk 2]
```diff
@@ -206,16 +206,34 @@ def rope_cache(self, q, k, v, qkv, position, fwd_ctx: ForwardContext):
...
-                    v,
-                    k_cache,
-                    v_cache,
-                    k_scale,
-                    v_scale,
-                    attn_metadata.slot_mapping,
-                    asm_layout=asm_layout,
-                )
+                if use_triton_attn:
+                    # pa_decode_gluon expects a global scalar scale (kv_quant_mode=0),
```
Comment (valarLip) [inline atom/model_ops/attention_mha.py]: i don;t think we need these as same swa for gptoss needed

**Review by valarLip — CHANGES_REQUESTED** [review]

File: atom/model_engine/model_runner.py (hunk 1 above)
Comment (wufann) [reply atom/model_engine/model_runner.py]: Fixed, calculate the memory of full attention and SWA separately.

File: atom/model_ops/attention_mha.py (hunk 2 above)
Comment (wufann) [reply atom/model_ops/attention_mha.py]: fixed. use 5D v_cache allocation to support per-token FP8 KV-cache quantization with asm_layout=True

**Review by valarLip — APPROVED** [review]
